src/test/java/com/neo4j/docker/TestUpgrade.java

```diff
@@ -4,60 +4,208 @@
+            db.putInitialDataIntoContainer( user, password );
+        }
+
+        try ( GenericContainer container = makeContainer( TestSettings.IMAGE_ID ) )
+        {
+            HostFileSystemOperations.mountHostFolderAsVolume( container, dataMount, "/data" );
+            container.start();
+            DatabaseIO db = new DatabaseIO( container );
+            db.verifyDataInContainer( user, password );
+        }
+    }
+
+    @Test
+    void canUpgradeFromReleasedVersion() throws Exception
+    {
+        var targetNeo4jVersion = TestSettings.NEO4J_VERSION;
+
+        // If this is the very first in a new major series (i.e. a .0.0 release) then this test isn't expected to work
+        Assumptions.assumeFalse( targetNeo4jVersion.major == 0 && targetNeo4jVersion.minor == 0 );
+
+        // TODO: update this when moving to the next minor release
+        // I am taking a guess here that we will have published 4.1.0 by 1st of July
+        if ( Instant.now().isBefore( Instant.parse( "2020-07-01T00:00:00.00Z" ) ) )
+        {
+            // This sort-of-hack is necessary when we cut a new minor branch before the previous minor branch has been published to dockerhub.
+            Assumptions.assumeTrue( Neo4jVersion.NEO4J_VERSION_420.isNewerThan( targetNeo4jVersion ) );
+        }
+        String fromImageName = dockerImageToUpgradeFrom( targetNeo4jVersion );
+
+        var testMountDirPrefix = String.format( "upgrade-from-%s", fromImageName ).replaceAll( ":", "_" );
+        Path testMountDir = HostFileSystemOperations.createTempFolder( testMountDirPrefix );
+
+        // TODO: test /plugins and /ssl directories
+        Map<String,@NotNull Path> readonlyMounts = createDirectories( testMountDir, List.of( "/conf" ) );
+        Map<String,@NotNull Path> writableMounts = createDirectories( testMountDir, List.of( "/data", "/logs", "/metrics" ) );
+
+        // write a value to neo4j.conf
+        try ( var confFile = new FileWriter( readonlyMounts.get( "/conf" ).resolve( "neo4j.conf" ).toFile() ) )
+        {
+            confFile.write( "dbms.memory.pagecache.size=8m" );
+        }
+
+        var allMounts = new HashMap<String,Path>();
+        allMounts.putAll( readonlyMounts );
+        allMounts.putAll( writableMounts );
+
+        try ( var container = makeContainer( fromImageName ) )
+        {
+            allMounts.forEach( ( containerMount, hostFolder ) -> HostFileSystemOperations.mountHostFolderAsVolume( container, hostFolder, containerMount ) );
+
+            final var startTime = Instant.now().toEpochMilli();
+            container.start();
+            DatabaseIO db = new DatabaseIO( container );
+            db.putInitialDataIntoContainer( user, password );
+
+            validateConfig( db );
+
+            // validate that writes actually went to the mounted directories
+            writableMounts.forEach( ( containerMount, hostFolder ) -> assertDirectoryModifiedSince( hostFolder, startTime ) );
```

Comment:
If I didn't include this check to check that mount paths worked. I include it  because _without checking_ if we changed how the mount paths worked (e.g. `/data` became `/databases`) this test would still pass (using `/data`) but it wouldn't be doing what I intended it to do which is persisting data on that mount.

Reply:
but there's already a check that the data has been persisted:
`db.verifyDataInContainer( user, password );` on line 146

Reply:
That doesn’t check that the data is in the mounted directory - just that it’s in neo4j

Reply:
In the second container, we don't write the data again before checking that the data is there. How else would the data be there unless it's reading from the mounted folder?

Reply:
I’m checking that subsequent writes succeed and go there.
admittedly perhaps a niche concern.
It’s not just the data folder though - it’s all the writable folders so logs etc as well.

Reply:
it's valid to check that we can still write to the database with the mounted folder since we might have introduced permissions changes (eg new uid!). I just don't think we need to query the file system to do that when we are *also* actually writing something to the database, since the database write will fail if something on the file system changed.

Reply:
I stand by this. The writing of metrics to `/metrics` is broken in 3.5 if the volume is mounted without permissions even when running as root. That is detected by this check. 

If I remove this check we do not detect that brokenness and this test passes when it should not.